transfer/basic_download.go

```diff
@@ -0,0 +1,240 @@
+	f, err := os.OpenFile(a.downloadFilename(t), os.O_RDWR, 0644)
+
+	if err != nil {
+		// Create a new file instead, must not already exist or error (permissions / race condition)
+		newfile, err := os.OpenFile(a.downloadFilename(t), os.O_CREATE|os.O_WRONLY|os.O_EXCL, 0644)
+		return newfile, 0, nil, err
+	}
+
+	// Successfully opened an existing file at this point
+	// Read any existing data into hash then return file handle at end
+	hash := tools.NewLfsContentHash()
+	n, err := io.Copy(hash, f)
+	if err != nil {
+		f.Close()
+		return nil, 0, nil, err
+	}
+	tracerx.Printf("xfer: Attempting to resume download of %q from byte %d", t.Object.Oid, n)
+	return f, n, hash, nil
+
+}
+
+// Create or open a download file for resuming
+func (a *basicDownloadAdapter) downloadFilename(t *Transfer) string {
+	// Not a temp file since we will be resuming it
+	return filepath.Join(a.tempDir(), t.Object.Oid+".tmp")
+}
+
+// download starts or resumes and download. Always closes dlFile if non-nil
+func (a *basicDownloadAdapter) download(t *Transfer, cb TransferProgressCallback, authOkFunc func(), dlFile *os.File, fromByte int64, hash hash.Hash) error {
+
+	if dlFile != nil {
+		// ensure we always close dlFile. Note that this does not conflict with the
+		// early close below, as close is idempotent.
+		defer dlFile.Close()
+	}
+
+	rel, ok := t.Object.Rel("download")
+	if !ok {
+		return errors.New("Object not found on the server.")
+	}
+
+	req, err := httputil.NewHttpRequest("GET", rel.Href, rel.Header)
+	if err != nil {
+		return err
+	}
+
+	if fromByte > 0 {
+		if dlFile == nil || hash == nil {
+			return fmt.Errorf("Cannot restart %v from %d without a file & hash", t.Object.Oid, fromByte)
+		}
+		// We could just use a start byte, but since we know the length be specific
+		req.Header.Set("Range", fmt.Sprintf("bytes=%d-%d", fromByte, t.Object.Size))
+	}
+
+	res, err := httputil.DoHttpRequest(req, true)
+	if err != nil {
+		// Special-case status code 416 () - fall back
+		if fromByte > 0 && dlFile != nil && res.StatusCode == 416 {
+			tracerx.Printf("xfer: server rejected resume download request for %q from byte %d; re-downloading from start", t.Object.Oid, fromByte)
+			dlFile.Close()
+			os.Remove(dlFile.Name())
+			return a.download(t, cb, authOkFunc, nil, 0, nil)
+		}
+		return errutil.NewRetriableError(err)
+	}
+	httputil.LogTransfer("lfs.data.download", res)
+	defer res.Body.Close()
+
+	// Range request must return 206 & content range to confirm
+	if fromByte > 0 {
+		rangeRequestOk := false
+		var failReason string
+		// check 206 and Content-Range, fall back if either not as expected
+		if res.StatusCode == 206 {
+			// Probably a successful range request, check Content-Range
+			if rangeHdr := res.Header.Get("Content-Range"); rangeHdr != "" {
+				regex := regexp.MustCompile(`bytes (\d+)\-.*`)
+				match := regex.FindStringSubmatch(rangeHdr)
+				if match != nil && len(match) > 1 {
+					contentStart, _ := strconv.ParseInt(match[1], 10, 32)
```

Comment:
I suspect this should be `ParseInt(match[1], 10, 64)` otherwise you can't resume when more than 2^31 bytes have already been downloaded